numlib/changeuidgid.c

```diff
@@ -78,27 +79,31 @@ uid_t libmail_getuid(const char *uname, gid_t *pw_gid)
 		bufsize = 16384;        /* Should be more than enough */
 	}
 
-	buf = malloc(bufsize);
-	if (buf == NULL)
-	{
-		perror("malloc");
-		exit(1);
-	}
-
-
-	errno=ENOENT;
+	do {
+		buf = malloc(bufsize);
+		if (buf == NULL)
+		{
+			perror("malloc");
+			exit(1);
+		}
 
-	getpwnam_r(p, &pwbuf, buf, bufsize, &pw);
+		s = getpwnam_r(p, &pwbuf, buf, bufsize, &pw);
+		if (s == ERANGE) {
+			free(buf);
+			bufsize += 1024;
+		}
+	} while (s == ERANGE && bufsize <= 65536);
 
-	free(buf);
+	free(p);
 
 	if (pw == 0)
 	{
-		free(p);
+		errno = s;
 		perror("getpwnam_r");
 		exit(1);
 	}
-	free(p);
+
+	free(buf);
```

Comment:
I don't see how. If getpwnam_r fails, `pw` will be 0 in addition to `buf` getting `free`d. `buf` will be allocated on the next iteration of the `while` loop. If the loop does not iterate it will take the perror/exit path, and won't reach this `free`.

This hinges on `getpwnam_r` documented behavior of always setting `pw` to 0 on any failure.

Reply:
Yes, you're correct. I was mistaken.